File: Daily/2492.minimum-score-of-a-path-between-two-cities.js
/*
 * @lc app=leetcode id=2492 lang=javascript
 *
 * [2492] Minimum Score of a Path Between Two Cities
 */

// @lc code=start
/**
 * @param {number} n
 * @param {number[][]} roads
 * @return {number}
 */
var minScore = function (n, roads) {
    /** @type {Map<number, number>[]} */
    const neighbors = [];
    for (let i = 1; i <= n; i++) {
        neighbors[i] = new Map();
    }

    for (const [city1, city2, distance] of roads) {
        neighbors[city1].set(city2, distance);
        neighbors[city2].set(city1, distance);
    }

    roads.sort(([, , distance1], [, , distance2]) => distance1 - distance2);

    /** @type {Map<string, boolean>} */
    const hasPathMemo = new Map();
    /** @type {Set<number>} */
    const visited = new Set();

    for (const [city1, city2, distance] of roads) {
        if (
            hasPath(neighbors, 1, city1, visited, hasPathMemo) &&
            hasPath(neighbors, n, city2, visited, hasPathMemo)
        ) {
            return distance;
        }
    }

    return -1;
};

/**
 *
 * @param {Map<number, number>[]} neighbors
 * @param {number} city1
 * @param {number} city2
 * @param {Set<number>} visited
 * @param {Map<string, boolean>} memo
 * @returns {boolean}
 */
function hasPath(neighbors, city1, city2, visited, memo) {
    if (city1 === city2) return true;

    if (city1 > city2) {
        [city1, city2] = [city2, city1];
    }

    const memoKey = `${city1}-${city2}`;
    if (memo.has(memoKey)) {
        // @ts-ignore
        return memo.get(memoKey);
    }

    const city1Neighbors = neighbors[city1];
    if (city1Neighbors.has(city2)) {
        memo.set(memoKey, true);
        return true;
    }

    for (const [city1Neighbor] of city1Neighbors) {
        if (!visited.has(city1Neighbor)) {
            visited.add(city1Neighbor);
            try {
                if (hasPath(neighbors, city1Neighbor, city2, visited, memo)) {
                    memo.set(memoKey, true);
                    return true;
                }
            } finally {
                visited.delete(city1Neighbor);
            }
        }
    }

    memo.set(memoKey, false);
    return false;
}
// @lc code=end

File: src/2026/2492.minimum-score-of-a-path-between-two-cities.cpp
/*
 * @lc app=leetcode id=2492 lang=cpp
 *
 * [2492] Minimum Score of a Path Between Two Cities
 */

#include <vector>

// @lc code=start

class UFSet {
 public:
  explicit UFSet(const int element_count) {
    parents.resize(element_count);
    for (int i = 0; i < parents.size(); i++) {
      parents[i] = i;
    }
  }

  void Union(const int element1, const int element2) {
    parents[Find(element2)] = parents[Find(element1)];
  }

  int Find(int element) {
    while (parents[element] != element) {
      parents[element] = parents[parents[element]];
      element = parents[element];
    }
    return element;
  }

  bool InOneSet(const int element1, const int element2) {
    return Find(element1) == Find(element2);
  }

 private:
  std::vector<int> parents;
};

class Solution {
 public:
  int minScore(const int n, const std::vector<std::vector<int>>& roads) {
    UFSet uf_set(n);

    for (const std::vector<int>& road : roads) {
      const int node1 = road[0] - 1;
      const int node2 = road[1] - 1;
      uf_set.Union(node1, node2);
    }

    int min_distance = INT_MAX;
    for (const std::vector<int>& road : roads) {
      const int node1 = road[0] - 1;
      const int distance = road[2];

      if (!uf_set.InOneSet(node1, 0)) {
        continue;
      }

      min_distance = std::min(min_distance, distance);
    }

    return min_distance;
  }
};
// @lc code=end
